Sources/PitchSpeller/Data Structures/_Graph.swift

```diff
@@ -0,0 +1,293 @@
+
+// MARK: - Weightedness Flags
+
+protocol Weightedness { }
+protocol Unweighted: Weightedness { }
+protocol AsWeight: Weightedness { }
+
+// Allows Double to be used as an edge weight
+extension Double: AsWeight { }
+enum WithoutWeights: Unweighted {
+    case unweighted
+}
+
+// Weightable, directable implementation of a Graph structure
+struct _Graph<Weight: Weightedness, Pair: SymmetricPair & Directedness & Hashable> where Pair.A: Hashable {
+    
+    // MARK: - Typealiases
+    
+    typealias Node = Pair.A
+    
+    struct Edge {
+        
+        // MARK: - Instance Properties
+        
+        let nodes: Pair
+        let weight: Weight
+        
+        // MARK: - Initializers
+        
+        init (_ a: _Graph.Node, _ b: _Graph.Node, withWeight weight: Weight) {
+            self.nodes = Pair(a, b)
+            self.weight = weight
+        }
+        
+        init (_ nodes: Pair, withWeight weight: Weight) {
+            self.nodes = nodes
+            self.weight = weight
+        }
+    }
+    
+    struct Path {
+        
+        // MARK: - Instance Properties
+        
+        let nodes: [Node]
+        let weights: [Pair: Weight]
+        
+        // MARK: - Initializers
+        
+        init (_ nodes: [Node], _ weights: [Pair: Weight]) {
+            self.nodes = nodes
+            self.weights = weights
+        }
+    }
+    
+    // MARK: - Instance Properties
+     
+    var nodes: Set<Node>
+    var adjacents: [Pair: Weight]
+    
+    var edges: [Edge] {
+        return adjacents.compactMap { Edge($0.key, withWeight: $0.value) }
+    }
+    
+    // MARK: - Instance Methods
+    
+    mutating func insertNode (_ node: Node) {
+        nodes.insert(node)
+    }
+    
+    mutating func insertEdge (from source: Node, to destination: Node, withWeight weight: Weight) {
+        adjacents[Pair(source, destination)] = weight
+    }
+    
+    mutating func insertEdge (_ keyValue: (Pair, Weight)) {
+        insertEdge(from: keyValue.0.a, to: keyValue.0.b, withWeight: keyValue.1)
+    }
+    
+    mutating func insertEdge(_ pair: Pair, _ weight: Weight) {
+        insertEdge(from: pair.a, to: pair.b, withWeight: weight)
+    }
+
+    mutating func updateEdge(_ pair: Pair, with transform: (Weight) -> Weight) {
+        guard let weight = weight(pair) else { return }
+        insertEdge(pair, transform(weight))
+    }
+    
+    mutating func insertPath (_ path: Path) {
+        path.nodes.forEach { insertNode($0) }
+        path.weights.forEach { insertEdge($0) }
+    }
+    
+    mutating func removeEdge (from source: Node, to destination: Node) {
+        adjacents[Pair(source, destination)] = nil
+    }
+    
+    /// - Returns: Weight of the edge from `source` to `destination` if it exists, else nil
+    func weight (from source: Node, to destination: Node) -> Weight? {
+        return adjacents[Pair(source, destination)]
+    }
+    
+    func weight (_ pair: Pair) -> Weight? {
```

Comment:
Super small nit, and probably another do-as-I-say-not-: higher-level funcs should be placed above the lower level func they call.

Reply:
👍